File: session-02-dependency-injection/java/src/main/java/coffeeshop/order/OrderController.java
package coffeeshop.order;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/orders")
public class OrderController {

    // The controller depends on OrderService, not OrderRepository.
    // Each layer declares only what it needs. The container wires the chain.
    private final OrderService orderService;

    public OrderController(OrderService orderService) {
        this.orderService = orderService;
    }

    @GetMapping
    public List<Order> listOrders() {
        return orderService.listOrders();
    }

    @GetMapping("/{id}")
    public ResponseEntity<Order> getOrder(@PathVariable int id) {
        return orderService.findOrder(id)
            .map(ResponseEntity::ok)
            .orElse(ResponseEntity.notFound().build());
    }
}

File: session-02-dependency-injection/typescript/src/order/order.controller.ts
import {Controller, Get, NotFoundException, Param, ParseIntPipe} from "@nestjs/common";
import {Order} from "./order.model";
import {OrderService} from "./order.service";

@Controller("orders")
export class OrderController {
    // The controller depends on OrderService, not OrderRepository.
    // Each layer declares only what it needs. NestJS wires the chain.
    constructor(private readonly orderService: OrderService) {
    }

    @Get()
    listOrders(): Order[] {
        return this.orderService.listOrders();
    }

    @Get(":id")
    getOrder(@Param("id", ParseIntPipe) id: number): Order {
        const order = this.orderService.findOrder(id);
        if (!order) {
            throw new NotFoundException(`Order ${id} not found`);
        }
        return order;
    }
}
